scripts-python/case.build

```diff
@@ -0,0 +1,508 @@
+    shared_lib = os.path.join(sharedpath, "lib")
+    shared_inc = os.path.join(sharedpath, "include")
+    for shared_item in [shared_lib, shared_inc]:
+        if (not os.path.exists(shared_item)):
+            os.makedirs(shared_item)
+
+    libs = ["mct", "gptl", "pio", "csm_share"]
+    logs = []
+
+    for lib in libs:
+        full_lib_path = os.path.join(sharedpath, lib)
+        if (not os.path.exists(full_lib_path)):
+            os.makedirs(full_lib_path)
+
+        file_build = os.path.join(sharedpath, "%s.bldlog.%s" % (lib, lid))
+        with open(file_build, "w") as fd:
+            fd.write("Current env:\n%s" % "\n".join(["  %s = %s" % (env, os.environ[env]) for env in sorted(os.environ)]))
+
+        my_file = os.path.join(os.path.dirname(machines_file), "buildlib.%s" % lib)
+        stat = run_cmd("%s %s %s >> %s 2>&1" % (my_file, sharedpath, caseroot, file_build), from_dir=exeroot, ok_to_fail=True)[0]
+        expect(stat == 0, "ERROR: buildlib.%s failed, cat %s" % (lib, file_build))
+        logs.append(file_build)
+
+    return logs
+
+###############################################################################
+def build_model(case, build_threaded, exeroot, clm_config_opts, incroot,
+                comp_atm,   comp_lnd,   comp_ice,   comp_ocn,   comp_glc,   comp_wav,   comp_rof,
+                nthrds_atm, nthrds_lnd, nthrds_ice, nthrds_ocn, nthrds_glc, nthrds_wav, nthrds_rof, lid, caseroot, cimeroot):
+###############################################################################
+    config_atm_dir = os.path.dirname(case.get_value("CONFIG_ATM_FILE"))
+    config_lnd_dir = os.path.dirname(case.get_value("CONFIG_LND_FILE"))
+    config_ice_dir = os.path.dirname(case.get_value("CONFIG_ICE_FILE"))
+    config_ocn_dir = os.path.dirname(case.get_value("CONFIG_OCN_FILE"))
+    config_glc_dir = os.path.dirname(case.get_value("CONFIG_GLC_FILE"))
+    config_wav_dir = os.path.dirname(case.get_value("CONFIG_WAV_FILE"))
+    config_rof_dir = os.path.dirname(case.get_value("CONFIG_ROF_FILE"))
+
+    # Also defines build order
+    models_build_data = [("atm", comp_atm, config_atm_dir, nthrds_atm),
+                         ("lnd", comp_lnd, config_lnd_dir, nthrds_lnd),
+                         ("ice", comp_ice, config_ice_dir, nthrds_ice),
+                         ("ocn", comp_ocn, config_ocn_dir, nthrds_ocn),
+                         ("glc", comp_glc, config_glc_dir, nthrds_glc),
+                         ("wav", comp_wav, config_wav_dir, nthrds_wav),
+                         ("rof", comp_rof, config_rof_dir, nthrds_rof)]
+
+    logs = []
+    overall_smp = os.environ["SMP"]
+
+    for model, comp, config_dir, nthrds in models_build_data:
+        os.environ["MODEL"] = model
+        if (nthrds > 1 or build_threaded == "TRUE"):
+            os.environ["SMP"] = "TRUE"
+        else:
+            os.environ["SMP"] = "FALSE"
+
+        # TODO - Special case for clm?
+        # TODO - bldroot?
```

Comment:
I'm trying to understand how bldroot works in this algorithm. It looks like it might be left as empty string in many cases in the perl version. Is this right?